Fix for #5058: collect_system_data_files() scans parent

rokm opened the pull request:
Fixes #5058 by removing the jump to the parent directory in `collect_system_data_files`. 

This changes the returned paths to the collected files - the paths are not prefixed by the last component of the input search path anymore. Therefore all hooks using `collect_system_data_files` need to be adjusted in order to avoid misplacing the files in the dist directory.

htgoebel commented:
Many thanks for tracking this down!

The offending `os.dirname` was introduced by my in 5d6715a24a51928658e38c26df596ca8ed7de386, trying to solve subtle difference on Windows (as the comment says), which are described in detail in #4707.

When applying this pull-request, this will change the API of `collect_system_data_files()`, maybe breaking hooks provided by external packages. Su I propose a different fix:
```
path = os.path.normpath(path)
# ensure there is a *single* os.sep at the end, required for `remove_prefix`
path = path.strip(os.sep) + os.sep
```

What do you think?

rokm commented:
> Many thanks for tracking this down!
> 
> The offending `os.dirname` was introduced by my in [5d6715a](https://github.com/pyinstaller/pyinstaller/commit/5d6715a24a51928658e38c26df596ca8ed7de386), trying to solve subtle difference on Windows (as the comment says), which are described in detail in #4707.
> 
> When applying this pull-request, this will change the API of `collect_system_data_files()`, maybe breaking hooks provided by external packages. Su I propose a different fix:
> 
> ```
> path = os.path.normpath(path)
> # Ensure `path` ends with a *single* os.sep, required for `remove_prefix`
> path = path.rstrip(os.sep) + os.sep
> ```
> 
> What do you think?

But won't the proposed fix also change the `collect_system_data_files()` API/behavior? Because once that jump to parent folder is fixed, the returned collected files will be relative to the give input path instead of its parent. And currently all our hooks seem to expect this behavior (path relative to parent of input path instead of input path itself) - hence I had to fix them in subsequent commits.

So I'm a bit puzzled at how things worked before the `os.dirname` was introduced in the first place...

htgoebel commented:
I'd say, 5d6715a24a51928658e38c26df596ca8ed7de386 introduced the bug. If you revert this commit (as of today this requires resolving a conflict, which is simple) the issue is gone.

Maybe we should try to get rid of doing path manipulation ourself and use pathlib instead. I scratched this in <https://github.com/htgoebel/pyinstaller/commit/3159f1ca8f900e215a2e5e6b3bfae300a916a66c>. (In the long run it might make sense to switch to pathlib for all path-related stuff, anyway).

rokm commented:
I took another look at 5d6715a24a51928658e38c26df596ca8ed7de386, and now I see that it introduced both the jump to parent (via the added `dirname`), but it also changed the way `dest` path is computed (by removing a `dirname` there). So on the whole, the returned paths remain unchanged - they were relative to the parent of the given input bath before that commit, and they remain such after it as well. (And that's why reverting the whole commit indeed solves the issue).

I think it makes more sense for the returned file paths to be relative to the given search path instead of its parent, though. But it indeed means that we need to fix our hooks, and that any external users of `collect_system_data_files` will need to fix theirs, too.

And now that I think about it, we might need to check the behavior of `collect_data_files` - because from what I see 5d6715a24a51928658e38c26df596ca8ed7de386, removed the `dirname` in computation of `dest` path, but did not introduce an additional `dirname` in search path, so the overall behavior might have changed...

rokm commented:
I gave  htgoebel@3159f1c a try, but it changes the returned file paths so that they are relative to search path, same as this PR. So the hooks using `collect_system_data_files` will need to be fixed, same as with this PR.

But if the base fix in htgoebel@3159f1c  is better than babfaafed4e6f277a0e2f9e252c4b32877753c08 used here, then I'm all for using that instead.

htgoebel commented:
Yes, we should have a look at collect_data_files, too. This is documented in the manual, thus changing the befaviour here means breaking other code (there might not be much of it, though). To avoid confusion, both functions should follow the same idea.

My primary point in https://github.com/htgoebel/pyinstaller/commit/3159f1c is about using pathlib. This one is just a scratch, I did not take care about the correct path level.

htgoebel commented:
@rokm Anything you need to finish this?

rokm commented:
@htgoebel No, it should be complete now. (The last commit is the same as the one in #5150, so I'll drop it in next update push).

I left it at draft because, as you mentioned earlier, it modifies the behavior (hence all the fixes to our hooks), which may break external users. So if you want to preserve the old behavior, then fixes in this PR are not really required.

Although to me, the new behavior makes more sense (i.e., returned paths being relative to the input path). and it is reflected in the target paths in our hooks, which now also make more sense (e.g., `PySide2` and `PySide2/Qt` vs. `.` and `Qt`).

Legorooj commented:
@rokm could you resolve the conflicts and fix the CI failure?

rokm commented:
I've rebased the PR. 

Not sure if the CI failure was related to the PR, though - I'll look into it if it fails again this time around.

bwoodsend commented:
CI has passed. Can we get this merged? It appears to be collecting related issues...

Legorooj commented:
@rokm thanks for this PR!
